atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/OracleTableNameUnmapper.java

```diff
@@ -15,34 +15,75 @@
  */
 package com.palantir.atlasdb.keyvalue.dbkvs;
 
+import java.sql.SQLException;
+import java.util.concurrent.ExecutionException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
 import com.google.common.collect.Iterables;
 import com.palantir.atlasdb.AtlasDbConstants;
 import com.palantir.atlasdb.keyvalue.api.TableReference;
 import com.palantir.atlasdb.keyvalue.dbkvs.impl.ConnectionSupplier;
 import com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs;
 import com.palantir.atlasdb.keyvalue.impl.TableMappingNotFoundException;
 import com.palantir.nexus.db.sql.AgnosticResultSet;
+import com.palantir.nexus.db.sql.SqlConnection;
 
 class OracleTableNameUnmapper {
+    private static final Logger log = LoggerFactory.getLogger(OracleTableNameUnmapper.class);
+
     private final ConnectionSupplier conns;
 
+    private LoadingCache<String, String> unmappingCache = CacheBuilder.newBuilder().build(
+            new CacheLoader<String, String>() {
+                @Override
+                public String load(String fullTableName) throws Exception {
+                    SqlConnection conn = null;
+                    try {
+                        conn = conns.getFresh();
+                        AgnosticResultSet results = conn.selectResultSetUnregisteredQuery(
+                                "SELECT short_table_name "
+                                        + "FROM " + AtlasDbConstants.ORACLE_NAME_MAPPING_TABLE
+                                        + " WHERE table_name = ?", fullTableName);
+                        if (results.size() == 0) {
+                            throw new TableMappingNotFoundException(
+                                    "The table " + fullTableName + " does not have a mapping."
+                                            + "This might be because the table does not exist.");
+                        }
+
+                        return Iterables.getOnlyElement(results.rows()).getString("short_table_name");
+                    } finally {
+                        if (conn != null) {
+                            try {
+                                conn.getUnderlyingConnection().close();
```

Comment:
we might be able to call `conns.close()` in `OracleTableNameGetter` like most `Dbkvs` calls:

```
    public String getInternalShortTableName() throws TableMappingNotFoundException {
        ConnectionSupplier conns = new ConnectionSupplier(connectionSupplier);
        if (useTableMapping) {
            try {
                return new OracleTableNameUnmapper(conns).getShortTableNameFromMappingTable(tablePrefix,
                        tableRef);
            } finally {
                conns.close();
            }
        }
        return getPrefixedTableName();
    }
```